skrub/_tuning.py

```diff
@@ -0,0 +1,1334 @@
+>>> np.percentile([unwrap(v) for v in n.rvs(1000, random_state=0)], q)
+array([ 1.,  3.,  9., 29., 99.])
+
+If we need floating-point numbers rather than ints we use ``choose_float``:
+
+>>> alpha = choose_float(.01, 100, log=True, name='α')
+>>> alpha
+choose_float(0.01, 100, log=True, name='α')
+>>> alpha.rvs() # doctest: +SKIP
+NumericOutcome(value=16.656593316727974, name=None, in_choice='α', is_from_log_scale=True)
+
+The default outcome of a numeric choice is the middle of its range (either on a
+linear or log scale):
+
+>>> unwrap_default(choose_float(0.0, 100.0, log=False))
+50.0
+>>> unwrap_default(choose_float(1.0, 100.0, log=True))
+10.000000000000002
+
+It is possible to specify a number of steps on the range of value to discretize
+it. Too big a step size loses the benefits of a randomized search. However,
+discretizing the range is useful to make better use of caching by reusing the
+same values. If we set a number of steps, we get a ``DiscretizedNumericChoice``.
+
+>>> n_dims = choose_int(10, 100, name='n dims', log=True, n_steps=5)
+>>> n_dims
+choose_int(10, 100, log=True, n_steps=5, name='n dims')
+>>> type(n_dims)
+<class 'skrub._tuning.DiscretizedNumericChoice'>
+
+Discretized ranges are sequences so they can be used either with
+``RandomizedSearchCV`` or ``GridSearchCV``.
+
+>>> n_dims.rvs() # doctest: +SKIP
+NumericOutcome(value=32, name=None, in_choice='n dims', is_from_log_scale=True)
+>>> list(n_dims)
+[10, 18, 32, 56, 100]
+
+The default outcome is as close as possible to the middle while respecting the
+steps:
+
+>>> unwrap_default(n_dims)
+32
+
+Optional
+--------
+
+Finally, it is common to choose between something and nothing. The last type of
+choice is a shorthand for a choice between a given value and ``None``:
+
+>>> from skrub._tuning import optional
+>>> feature_selection = optional(SelectKBest(), name='feature selection')
+>>> feature_selection
+optional(SelectKBest(), name='feature selection')
+>>> type(feature_selection)
+<class 'skrub._tuning.Optional'>
+>>> list(feature_selection)
+[Outcome(value=SelectKBest(), name='true', in_choice='feature selection'), Outcome(value=None, name='false', in_choice='feature selection')]
+>>> unwrap_default(feature_selection)
+SelectKBest()
+"""  # noqa: E501
+
+import dataclasses
+import io
+import typing
+from collections.abc import Sequence
+
+import numpy as np
+from scipy import stats
+from sklearn.base import clone
+from sklearn.utils import check_random_state
+
+from . import _utils
+
+#
+# Choices and Outcomes
+# ====================
+#
+# We start by defining the different kinds of choices and their outcomes, and
+# the functions that construct them such as ``choose_from`` and
+# ``choose_float``.
+#
+
+
+def _with_fields(obj, **fields):
+    """
+    Make a copy of a dataclass instance with different values for some of the attributes
+    """
+    return obj.__class__(
+        **({f.name: getattr(obj, f.name) for f in dataclasses.fields(obj)} | fields)
```

Comment:
Why not:
```suggestion
from dataclasses import asdict
        **(asdict(obj) | fields)
```
?

Reply:
asdict recurses into attributes and makes a deep copy, here we want a shallow copy